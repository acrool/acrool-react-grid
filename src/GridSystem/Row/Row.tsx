'use client';

import React from 'react';
import styled, {css} from 'styled-components';
import {TStyledProps, IRowProps} from '../../types';
import {themeName} from '../../config';


/**
 * Row Component
 * align-items 預設加上 flex-start, 會讓鄰居Col高度不會一致
 * ps: 設定會 width 100% 會產生 margin 抵銷失敗
 */
const Row = styled.div<TStyledProps<IRowProps>>`
    --bear-gutter-x: ${props => props.theme[themeName]?.gridGutterWidth ?? '.625rem'};
    --bear-gutter-y: 0;
    box-sizing: border-box;
    display: flex;
    flex-wrap: wrap;
    margin-top: calc(var(--bear-gutter-y) * -1);
    margin-right: calc(var(--bear-gutter-x) * -0.5);
    margin-left: calc(var(--bear-gutter-x) * -0.5);
    padding-inline-start: 0; /* 避免 ul 預設樣式位移 */

    > * {
        flex-shrink: 0;
        width: 100%;
        max-width: 100%;
        padding-right: calc(var(--bear-gutter-x) * 0.5);
        padding-left: calc(var(--bear-gutter-x) * 0.5);
        margin-top: var(--bear-gutter-y);
    }

`;

export default Row;

