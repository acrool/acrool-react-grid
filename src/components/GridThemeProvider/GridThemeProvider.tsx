import React, {PropsWithChildren} from 'react';
import {ThemeProvider, createGlobalStyle} from 'styled-components';

import {defaultGridTheme, themeName} from '../../config';
import {IGridSetting} from '../../types';


interface IProps {
    gridTheme: Partial<IGridSetting>;
    children: React.ReactNode;
}

const ThemeProviderFixed = ThemeProvider as unknown as React.FC<PropsWithChildren<{ theme: any }>>;


/**
 * Grid Theme Provider
 * use :
 *     <GridThemeProvider gridTheme={{
 *         gridGutterWidth: 5,
 *     }}>
 *          <Container>, <Row>, <Col> ...then eat this setting
 *     </GridThemeProvider>
 */
const GridThemeProvider = (props: IProps) => {
    const {gridTheme, children} = props;
    const {gridBreakpoints, containerMaxWidths, gridGutterWidthMedia, ...baseDefaultTheme} = defaultGridTheme;

    // console.log('test render bear-react-grid version 10');
    const composeGridTheme = {
        [themeName]: {
            ...baseDefaultTheme,
            ...gridTheme,
            gridBreakpoints: Object.assign({}, gridBreakpoints, gridTheme?.gridBreakpoints),
            containerMaxWidths: Object.assign({}, containerMaxWidths, gridTheme?.containerMaxWidths),
            gridGutterWidthMedia: Object.assign({}, gridGutterWidthMedia, gridTheme?.gridGutterWidthMedia),
        }
    };

    return <ThemeProviderFixed theme={composeGridTheme}>
        {children}
        <GlobalStyled/>
    </ThemeProviderFixed>;
};

export default GridThemeProvider;


const GlobalStyled = createGlobalStyle`
    
`;
