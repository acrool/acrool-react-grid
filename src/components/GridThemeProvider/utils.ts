import {TStyledProps, IRowProps, NoXsMediaSize} from '../../types';
import {mediaSizes} from '../../config';
import media from '../../media';


const renderMarginStyle = (sizeName: NoXsMediaSize) => {
    return Array.from({length: 6}).map((row,idx) => {
        return `
        .m-${sizeName}-${idx} {margin: var(--bear-spacer-${idx}) !important;}
        .mt-${sizeName}-${idx}, .my-${sizeName}-${idx} {margin-top: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .mr-${sizeName}-${idx}, .mx-${sizeName}-${idx} {margin-right: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .mb-${sizeName}-${idx}, .my-${sizeName}-${idx} {margin-bottom: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .ml-${sizeName}-${idx}, .mx-${sizeName}-${idx} {margin-left: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        `;
    }).join('');
};

const renderPaddingStyle = (sizeName: NoXsMediaSize) => {
    return Array.from({length: 6}).map((row,idx) => {
        return `
        .p-${sizeName}-${idx} {margin: var(--bear-spacer-${idx}) !important;}
        .pt-${sizeName}-${idx}, .py-${sizeName}-${idx} {padding-top: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .pr-${sizeName}-${idx}, .px-${sizeName}-${idx} {padding-right: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .pb-${sizeName}-${idx}, .py-${sizeName}-${idx} {padding-bottom: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        .pl-${sizeName}-${idx}, .px-${sizeName}-${idx} {padding-left: ${idx === 0 ? 0: `var(--bear-spacer-${idx})`} !important;}
        `;
    }).join('');
};

const renderOrderStyle = (sizeName: NoXsMediaSize) => {
    return Array.from({length: 6}).map((row,idx) => {
        return `
        .order-${sizeName}-${idx} {order: ${idx}!important;}
        `;
    }).join('');
};


/**
 * 產生 RWD 樣式
 * @param props
 */
export const generateRWDStyled = (props: TStyledProps<IRowProps>) => {
    return mediaSizes
        .map(sizeName => {
            return media[sizeName]`
            .d-${sizeName}-none {display: none !important;}
            .d-${sizeName}-inline {display: inline !important;}
            .d-${sizeName}-inline-block {display: inline-block !important;}
            .d-${sizeName}-block {display: block !important;}
            .d-${sizeName}-flex {display: flex !important;}
            .d-${sizeName}-inline-flex {display: inline-flex !important;}
            .d-${sizeName}-table-cell {display: table-cell !important;}
            .d-${sizeName}-table {display: table !important;}
            .d-${sizeName}-table-row {display: table-row !important;}

            .flex-${sizeName}-row {flex-direction: row !important;}
            .flex-${sizeName}-column {flex-direction: column !important;}
            .flex-${sizeName}-row-reverse {flex-direction: row-reverse !important;}
            .flex-${sizeName}-column-reverse {flex-direction: column-reverse !important;}

            .flex-${sizeName}-wrap {flex-wrap: wrap !important;}
            .flex-${sizeName}-nowrap {flex-nowrap: nowrap !important;}
            .flex-${sizeName}-wrap-reverse {flex-nowrap: wrap-reverse !important;}

            .flex-${sizeName}-fill {flex: 1 1 auto !important;}
            .flex-${sizeName}-grow-0 {flex-grow: 0 !important;}
            .flex-${sizeName}-grow-1 {flex-grow: 1 !important;}
            .flex-${sizeName}-shrink-0 {flex-shrink: 0 !important;}
            .flex-${sizeName}-shrink-1 {flex-shrink: 1 !important;}

            .justify-content-${sizeName}-start {justify-content: flex-start !important;}
            .justify-content-${sizeName}-end {justify-content: flex-end !important;}
            .justify-content-${sizeName}-center {justify-content: flex-center !important;}
            .justify-content-${sizeName}-between {justify-content: space-between !important;}
            .justify-content-${sizeName}-around {justify-content: space-around !important;}

            .align-items-${sizeName}-start {align-items: flex-start !important;}
            .align-items-${sizeName}-end {align-items: flex-end !important;}
            .align-items-${sizeName}-center {align-items: center !important;}
            .align-items-${sizeName}-baseline {align-items: baseline !important;}
            .align-items-${sizeName}-stretch {align-items: stretch !important;}

            .align-content-${sizeName}-start {align-content: start !important;}
            .align-content-${sizeName}-end {align-content: flex-end !important;}
            .align-content-${sizeName}-center {align-content: center !important;}
            .align-content-${sizeName}-between {align-content: between !important;}
            .align-content-${sizeName}-around {align-content: around !important;}
            .align-content-${sizeName}-stretch {align-content: stretch !important;}

            .align-self-${sizeName}-start {align-self: flex-start !important;}
            .align-self-${sizeName}-end {align-self: flex-end !important;}
            .align-self-${sizeName}-center {align-self: center !important;}
            .align-self-${sizeName}-baseline {align-self: baseline !important;}
            .align-self-${sizeName}-stretch {align-self: stretch !important;}

            .text-${sizeName}-justify {text-align: justify !important;}
            .text-${sizeName}-left {text-align: left !important;}
            .text-${sizeName}-right {text-align: right !important;}
            .text-${sizeName}-center {text-align: center !important;}

            .m-${sizeName}-auto {margin: var(--bear-spacer-auto) !important;}
            .mt-${sizeName}-auto, .my-${sizeName}-auto {margin-top: auto !important;}
            .mr-${sizeName}-auto, .mx-${sizeName}-auto {margin-right: auto !important;}
            .mb-${sizeName}-auto, .my-${sizeName}-auto {margin-bottom: auto !important;}
            .ml-${sizeName}-auto, .mx-${sizeName}-auto {margin-left: auto !important;}

            ${renderMarginStyle(sizeName)}
            ${renderPaddingStyle(sizeName)}
            ${renderOrderStyle(sizeName)}

            .overflow-x-${sizeName}-hidden {overflow-x: hidden!important;}
            .overflow-y-${sizeName}-hidden {overflow-y: hidden!important;}
            .overflow-x-${sizeName}-auto {overflow-x: auto!important;}
            .overflow-y-${sizeName}-auto {overflow-y: auto!important;}
        `;
        });
};

