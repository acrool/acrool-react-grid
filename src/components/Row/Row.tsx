import React from 'react';
import styled, {css} from 'styled-components';
import media from '../../media';
import {FCChildrenProps, TStyledProps} from '../../types';
import {mediaSizes, themeName} from '../../config';



export interface IProps extends FCChildrenProps{
    noGutters?: boolean;
    vertical?: 'top'|'center'|'bottom';
    horizontal?: 'left'|'center'|'right';
}


/**
 * 產生 RWD 樣式
 * @param props
 */
const generateRWDStyled = (props: TStyledProps<IProps>) => {
    return mediaSizes
        .map(sizeName => {
            return media[sizeName]`
            --bs-gutter-x: ${props.theme[themeName]?.gridGutterWidthMedia[sizeName]};
            --bs-gutter-y: 0;
        `;
        });
};


/**
 * Row Component
 * align-items 預設加上 flex-start, 會讓鄰居Col高度不會一致
 * ps: 設定會 width 100% 會產生 margin 抵銷失敗
 */
const Row = styled.div.attrs((props: TStyledProps<IProps>) => ({
    'data-row': props.noGutters ? 'no-gutter': '',
    'data-horizontal': props.horizontal ? props.horizontal: undefined,
    'data-vertical': props.vertical ? props.vertical: undefined,
}))`
  ${(props: TStyledProps<IProps>) => css`
     ${!props.noGutters && css`
       ${generateRWDStyled(props)}
     `}
 `}
`;

export default Row;

