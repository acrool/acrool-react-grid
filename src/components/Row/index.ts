export {default} from './Row';
export {default as Flex} from './Flex';
