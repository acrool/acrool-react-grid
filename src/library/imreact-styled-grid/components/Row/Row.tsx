import React from 'react';
import styled, {css} from 'styled-components/macro';
import getDataName from './getDataName';
import {themeName} from '../../config';
import {RowProps} from './types';
import media from '../../media';

/**
 * Get Row Margin
 * @param props
 * @returns {*}
 */
const getRowMargin = (props: any) => (props.noGutters ? 0 : `-${props.theme.imreactStyledGrid.gridGutterWidth}px`);

const generateMedia = (props: any) => Object.keys(props.theme.imreactStyledGrid.gridBreakpoints)
    .map(sizeName => {
        if (props.theme.imreactStyledGrid.gridBreakpoints[sizeName] > 0) {
            return media[sizeName]`
                margin-left: -${props.theme.imreactStyledGrid.gridGutterWidthMedia[sizeName]}px;
                margin-right: -${props.theme.imreactStyledGrid.gridGutterWidthMedia[sizeName]}px;
            `;
        }
        return false;
    })
    .filter(row => row !== false);
;

/**
 * Row Component
 */
const Row: any = styled.div.attrs((props: any) => ({
    'data-grid': 'row',
    'data-debug': getDataName(props),
}))`
  display: -ms-flexbox;
  display: flex;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;

  ${(props: any) => css`
     margin-right: ${getRowMargin(props)};
     margin-left: ${getRowMargin(props)};

     // ps: 設定會 width 100% 會產生 margin 抵銷失敗


     ${!props.noGutters && css`
        ${generateMedia(props)}
     `}

     ${props.noGutters && css`
        >[data-grid=col]{
            padding-right: 0;
            padding-left: 0;
        }
     `}
 `}
`;

export default ({forwardAs, ...props}: RowProps) => <Row as={forwardAs} {...props}/>;

