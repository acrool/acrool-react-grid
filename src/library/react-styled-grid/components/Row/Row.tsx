import React from 'react';
import styled, {css} from 'styled-components/macro';
import getDataName from './getDataName';
import {themeName} from '../../config';
import {RowProps} from './types';
import media from '../../media';

/**
 * Get Row Margin
 * @param props
 * @returns {*}
 */
const getRowMargin = (props: any) => (props.noGutters ? 0 : `-${props.theme.reactStyledGrid.gridGutterWidth}px`);

const generateMedia = (props: any) => Object.keys(props.theme.reactStyledGrid.gridBreakpoints)
    .map(sizeName => {
        if (props.theme.reactStyledGrid.gridBreakpoints[sizeName] > 0) {
            return media[sizeName]`
                margin-left: -${props.theme.reactStyledGrid.gridGutterWidthMedia[sizeName]}px;
                margin-right: -${props.theme.reactStyledGrid.gridGutterWidthMedia[sizeName]}px;
            `;
        }
        return false;
    })
    .filter(row => row !== false);
;

/**
 * Row Component
 */
const Row: any = styled.div.attrs((props: any) => ({
    'data-grid': 'row',
    'data-debug': getDataName(props),
}))`
  display: -ms-flexbox;
  display: flex;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;

  ${(props: any) => css`
     margin-right: ${getRowMargin(props)};
     margin-left: ${getRowMargin(props)};

     // ps: 設定會 width 100% 會產生 margin 抵銷失敗


     ${!props.noGutters && css`
        ${generateMedia(props)}
     `}

     ${props.noGutters && css`
        >[data-grid=col]{
            padding-right: 0;
            padding-left: 0;
        }
     `}
 `}
`;

export default ({forwardAs, ...props}: RowProps) => <Row as={forwardAs} {...props}/>;

