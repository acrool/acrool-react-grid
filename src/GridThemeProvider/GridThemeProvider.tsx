import React, {PropsWithChildren} from 'react';
import {ThemeProvider, createGlobalStyle, css} from 'styled-components';

import {defaultGridTheme, themeName} from '../config';
import {TStyledProps, TThemeProps, IGridThemeProviderProps} from '../types';
import {generateRWDStyled} from './utils';
import '../styles.css';



const ThemeProviderFixed = ThemeProvider as unknown as React.FC<PropsWithChildren<{ theme: any }>>;


/**
 * Grid Theme Provider
 * use :
 *     <GridThemeProvider gridTheme={{
 *         gridGutterWidth: 5,
 *     }}>
 *          <Container>, <Row>, <Col> ...then eat this setting
 *     </GridThemeProvider>
 */
const GridThemeProvider = (props: IGridThemeProviderProps) => {
    const {gridTheme, children} = props;
    const {gridBreakpoints, containerMaxWidths, gridGutterWidthMedia, ...baseDefaultTheme} = defaultGridTheme;

    // console.log('test render bear-react-grid version 10');
    const composeGridTheme = {
        [themeName]: {
            ...baseDefaultTheme,
            ...gridTheme,
            gridBreakpoints: Object.assign({}, gridBreakpoints, gridTheme?.gridBreakpoints),
            containerMaxWidths: Object.assign({}, containerMaxWidths, gridTheme?.containerMaxWidths),
            gridGutterWidthMedia: Object.assign({}, gridGutterWidthMedia, gridTheme?.gridGutterWidthMedia),
        }
    };

    return <ThemeProviderFixed theme={composeGridTheme}>
        {children}
        <GlobalStyled/>
    </ThemeProviderFixed>;
};

export default GridThemeProvider;


const GlobalStyled = createGlobalStyle`
    :root{
        --bear-spacer: 1rem;
        --bear-spacer-1: calc(1rem * .25);
        --bear-spacer-2: calc(1rem * .5);
        --bear-spacer-3: calc(1rem);
        --bear-spacer-4: calc(1rem * 1.5);
        --bear-spacer-5: calc(1rem * 3);
    }

    ${(props: TStyledProps<TThemeProps>) => css`
        ${generateRWDStyled(props)}
    `}

    [data-row][data-horizontal="left"]{
        justify-content: flex-start;
    }
    [data-row][data-horizontal="center"]{
        justify-content: center;
    }
    [data-row][data-horizontal="right"]{
        justify-content: flex-end;
    }
    [data-row][data-vertical="top"]{
        justify-content: flex-start;
    }
    [data-row][data-vertical="center"]{
        justify-content: center;
    }
    [data-row][data-vertical="bottom"]{
        justify-content: flex-end;
    }
`;
