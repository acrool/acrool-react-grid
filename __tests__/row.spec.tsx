import React from 'react';
import 'jest-styled-components';
import {render} from '@testing-library/react';
import '@testing-library/dom';
import GridThemeProvider from '../src/GridThemeProvider';
import Row from '../src/GridSystem/Row';

describe('Row', () => {
    it('renders row', () => {
        const {container} = render(<Row>row</Row>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });

});
