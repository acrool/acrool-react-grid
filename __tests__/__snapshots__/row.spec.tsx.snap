// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row renders row 1`] = `
.c0 {
  --bear-gutter-x: 1.5rem;
  --bear-gutter-y: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-top: calc(var(--bear-gutter-y) * -1);
  margin-right: calc(var(--bear-gutter-x) * -0.5);
  margin-left: calc(var(--bear-gutter-x) * -0.5);
  padding-inline-start: 0;
}

.c0 > * {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 100%;
  max-width: 100%;
  padding-right: calc(var(--bear-gutter-x) * 0.5);
  padding-left: calc(var(--bear-gutter-x) * 0.5);
  margin-top: var(--bear-gutter-y);
}

<div
  class="c0"
  data-grid="row"
>
  row
</div>
`;
